src/inmanta/ast/__init__.py

```diff
@@ -465,6 +467,29 @@ def get_namespace(self) -> "Namespace":
     def get_location(self) -> Location:
         return self.location
 
+    def warn_shadowed_variables(self, nested_block: Optional["BasicBlock"] = None) -> None:
+        """
+            Produces a warning for any shadowed variables in ocurring in this namespace. This namespace's scope's root block
+            is used as an entrypoint for the check. If nested_block is provided, that block is interpreted as living in this
+            scope and only that block is searched for shadowing with respect to the scope.
+        """
+        assert self.scope is not None
+        shadowed_variables: Iterator[
+            Tuple[str, FrozenSet[Locatable], FrozenSet[Locatable]]
+        ] = self.scope.block.shadowed_variables(nested_blocks=[nested_block] if nested_block is not None else None)
```

Comment:
As it is now, this code is completely block bound, so I would like to see it in the block. 

It would perhaps make sense to put all declaration statements in the namespace. 
Then this would be more clear.

In a similar line perhaps the namespace should have a reference to the block (instead of going via the Scope)

Reply:
I took inspiration from this comment but didn't follow it to the letter:
- A block now keeps a list of it's definition statements. These are registered to the root block for each namespace during the building of the ast.
- `warn_shadowed_variables()` has been moved from `Namespace` to `BasicBlock`
Is this satisfactory?